lib/journald/src/lib.rs

```diff
@@ -0,0 +1,113 @@
+#![allow(non_camel_case_types)]
+
+#[macro_use]
+extern crate dlopen_derive;
+
+use dlopen::wrapper::{Container, WrapperApi};
+use libc::size_t;
+use std::collections::HashMap;
+use std::io::{Error as IOError, Result as IOResult};
+use std::iter;
+use std::os::raw::c_int;
+use std::ptr::null_mut;
+
+pub const SD_JOURNAL_LOCAL_ONLY: c_int = 1;
+pub const SD_JOURNAL_RUNTIME_ONLY: c_int = 2;
+pub const SD_JOURNAL_SYSTEM: c_int = 4;
+pub const SD_JOURNAL_CURRENT_USER: c_int = 8;
+
+pub enum sd_journal {}
+
+#[derive(WrapperApi)]
+struct LibSystemd {
+    sd_journal_open: extern "C" fn(ret: *mut *mut sd_journal, flags: c_int) -> c_int,
+    sd_journal_close: extern "C" fn(j: *mut sd_journal),
+    sd_journal_next: extern "C" fn(j: *mut sd_journal) -> c_int,
+    sd_journal_seek_head: extern "C" fn(j: *mut sd_journal) -> c_int,
+    sd_journal_restart_data: extern "C" fn(j: *mut sd_journal),
+    sd_journal_enumerate_data:
+        extern "C" fn(j: *mut sd_journal, data: *const *mut u8, l: *mut size_t) -> c_int,
+}
+
+fn load_lib() -> Result<Container<LibSystemd>, dlopen::Error> {
+    unsafe { Container::load("libsystemd.so") }
+}
+
+pub struct Journal {
+    lib: Container<LibSystemd>,
+    journal: *mut sd_journal,
+}
+
+unsafe impl Send for Journal {}
+
+pub type Record = HashMap<String, String>;
+
+impl Journal {
+    pub fn open(flags: c_int) -> IOResult<Journal> {
+        // Each Journal structure gets their own handle to the library,
+        // but I couldn't figure out how to make lazy_static work.
+        let lib = load_lib().map_err(|err| IOError::new(std::io::ErrorKind::Other, err))?;
```

Comment:
does this make sense to return an `io::Error` or to have a custom error enum?

Reply:
For everything other than `open`, the error value returned is copied from the libsystemd return, which is just a negative `errno`, so `io::Error` is appropriate. Here, it may also be a `dlopen::Error`, which I wrapped as `ErrorKind::Other`. If you think a new error enum is more appropriate than what this is doing, I'm open to that.

Reply:
I think specifically in this case I think it makes sense to provide an error enum so that its very easy to match on the dlopen::Error case? What do you think?

Reply:
Sure, I can do that.